src/connection.rs

```diff
@@ -17,484 +17,505 @@
             is_dead: false,
-            resource: Framed::new(resource, FrameCodec::new()),
             config,
-            id_counter: match mode {
+            streams: BTreeMap::new(),
+            resource: Framed::new(resource, FrameCodec::new()).fuse(),
+            incoming: VecDeque::new(),
+            pending: VecDeque::new(),
+            tasks: Vec::new(),
+            next_id: match mode {
                 Mode::Client => 1,
                 Mode::Server => 2
-            },
-            streams: BTreeMap::new(),
-            controller,
-            stream_rx,
-            stream_tx,
-            pending: None
+            }
         }
     }
 
-    /// Get a control handle which allows to open new streams.
-    pub fn control(&self) -> Ctrl {
-        self.controller.sender.clone()
+    fn next_stream_id(&mut self) -> Result<stream::Id, ConnectionError> {
+        if self.next_id > u32::MAX - 2 {
+            return Err(ConnectionError::NoMoreStreamIds)
+        }
+        let proposed = stream::Id::new(self.next_id);
+        self.next_id += 2;
+        match self.mode {
+            Mode::Client => assert!(proposed.is_client()),
+            Mode::Server => assert!(proposed.is_server())
+        }
+        Ok(proposed)
     }
 
-    fn open_stream(&mut self, data: Option<Body>) -> Result<(Stream, Frame<Data>), ConnectionError> {
-        trace!("open stream");
-        let id = self.next_stream_id()?;
-        let credit = self.config.receive_window;
-        let stream = self.new_stream(id, false, credit);
-        let mut frame = Frame::data(id, data.unwrap_or_else(Body::empty));
-        frame.header_mut().syn();
-        Ok((stream, frame))
+    fn is_valid_remote_id(&self, id: stream::Id, ty: Type) -> bool {
+        match ty {
+            Type::Ping | Type::GoAway => return id.is_session(),
+            _ => {}
+        }
+        match self.mode {
+            Mode::Client => id.is_server(),
+            Mode::Server => id.is_client()
+        }
     }
 
-    fn on_stream_item(&mut self, item: (stream::Id, Item)) -> RawFrame {
-        match item.1 {
-            Item::Data(body) => {
-                let mut frame = Frame::data(item.0, body);
-                if let Some(stream) = self.streams.get_mut(&item.0) {
-                    if stream.ack {
-                        stream.ack = false;
-                        frame.header_mut().ack()
-                    }
-                }
-                frame.into_raw()
+    fn wake_tasks(&mut self) {
+        for task in self.tasks.drain(..) {
+            task.notify()
+        }
+    }
+
+    fn flush_pending(&mut self) -> Poll<(), ConnectionError> {
+        while let Some(frame) = self.pending.pop_front() {
+            trace!("{:?}: send: {:?}", self.mode, frame.header);
+            if let AsyncSink::NotReady(frame) = self.resource.start_send(frame)? {
+                self.pending.push_front(frame);
+                try_ready!(self.resource.poll_complete())
```

Comment:
I think this line should be `return Async::NotReady;`.

The documentation of `start_send` mentions that calling `poll_complete()` here is useless:
> The method returns AsyncSink::NotReady if the sink was unable to begin sending, usually due to being full. The sink must have attempted to complete processing any outstanding requests (equivalent to poll_complete) before yielding this result.

Also, I guess there's a possibility that `start_send` returns `NotReady` and `poll_complete()` returns `Ready`, in which case this will loop forever.